File: Nim/Benchmark/src/benchmark.nim
# benchmark.nim
# Some (micro-)benchmarks for the Nim programming language
#
# vim: ft=nim sw=4 ts=4 et
#

from times import cpuTime
from bigints import toString

import fib
import perfect_numbers
import mandelbrot
from mandelbrot_colormap import Color

var
    tic : float


# Fibonacci numbers
# =================
echo("Fibonacci numbers:")
echo("---------------------")

tic = cpuTime()
echo("fib_naive(35) = ", fib_naive(35), "\tElapsed: ", cpuTime()-tic, "s")

tic = cpuTime()
echo("fib(35) = ", fib(35).toString(), "\tElapsed: ", cpuTime()-tic, "s")

tic = cpuTime()
echo("fib(1000) = ", fib(1000).toString(), " Elapsed: ", cpuTime()-tic, "s")

tic = cpuTime()
echo("fib_iter(35) = ", fib_iter(35).toString(), "\tElapsed: ", cpuTime()-tic, "s")

tic = cpuTime()
echo("fib_iter(1000) = ", fib_iter(1000).toString(), "\tElapsed: ", cpuTime()-tic, "s")

echo()

# Perfect numbers
# ===============
echo("Perfect numbers:")
echo("----------------")

tic = cpuTime()
echo("perfect_numbers(10000) = ", perfect_numbers(10000), "\tElapsed: ", cpuTime()-tic, "s")

#tic = cpuTime()
#echo("perfect_numbers_2(10000) = ", perfect_numbers_2(10000), "\tElapsed: ", cpuTime()-tic, "s")

echo()

# Mandelbrot sets
# ===============
echo("Mandelbrot sets:")
echo("----------------")

tic = cpuTime()
var image_small : array[180*120, Color]
discard mandelbrot(180, 120, -0.5, 0.0, 4.0/180, image_small)
echo("mandelbrot set (180 x 120)\tElapsed: ", cpuTime()-tic, "s")

tic = cpuTime()
discard write_pgm("image_small.pgm", image_small, 180, 120)
echo("image (180 x 120) written to file\tElapsed: ", cpuTime()-tic, "s")

tic = cpuTime()
var image_large : array[640*480, Color]
discard mandelbrot(640, 480, -0.5, 0.0, 4.0/640, image_large)
echo("mandelbrot set (640 x 480)\tElapsed: ", cpuTime()-tic, "s")

tic = cpuTime()
discard write_pgm("image_large.pgm", image_large, 640, 480)
echo("image (640 x 480) written to file\tElapsed: ", cpuTime()-tic, "s")

echo()

echo("---------------------")
echo()
echo("Done!")

echo("Press ENTER to continue..")
discard stdin.readLine()

File: JS/Benchmark/benchmark.js
// benchmark.js
// Some (micro-)benchmarks for the JavaScript (node/io) programming language
//
// vim: ft=javascript sw=4 ts=4
//

'use strict';

const fib = require("./fib");
const BigInt = require('BigInt');
const pn = require("./perfect_numbers");
const mandel = require("./mandelbrot");

console.log("JavaScript benchmarks");
console.log("=====================");
console.log();

console.log("Fibonacci numbers:");
console.log("------------------");

var tic = new Date();
var res = fib.fib_naive(35);
var toc = new Date();
console.log("fib_naive(35) = ", res, "\tElapsed: ", (toc.getTime()-tic.getTime()), "ms.");

var tic = new Date();
var fib_promise = fib.fib_naive_async(35);
fib_promise.then((res) => {
  var toc = new Date();
  console.log("fib_naive_async(35) = ", res, "\tElapsed: ", (toc.getTime()-tic.getTime()), "ms.");
});

tic = new Date();
res = fib.fib(35);
toc = new Date();
console.log("fib(35) = ", BigInt.bigInt2str(res, 10), "\tElapsed: ", (toc.getTime()-tic.getTime()), "ms.");

tic = new Date();
res = fib.fib(1000);
toc = new Date();
console.log("fib(1000) = ", BigInt.bigInt2str(res, 10), "\tElapsed: ", (toc.getTime()-tic.getTime()), "ms.");
console.log();

console.log("Perfect numbers:");
console.log("----------------");

tic = new Date();
res = pn.perfect_numbers(10000);
toc = new Date();
console.log("perfect_numbers(10000) = ", res, "\tElapsed: ", (toc.getTime()-tic.getTime()), "ms.");
tic = new Date();
res = [];
// the following version seems not to be completely lazy:
//for (let n of pn.sequence()) {
//   	if (n < 1000) res.push(n); else break;
//}
var gen = pn.sequence();
for (let i=0; i<4; i++) {
   	res.push(gen.next().value);
}
toc = new Date();
console.log("4 perfect numbers (Generator) = ", res, "\tElapsed: ", (toc.getTime()-tic.getTime()), "ms.");
console.log();


console.log("Mandelbrot set:");
console.log("--------------");

var tic11 = new Date();
let image = mandel.mandelbrot(640, 480, -0.5, 0.0, 4.0/640);
var toc11 = new Date();
console.log("Mandelbrot set (640x480) calculated!", "\tElapsed: ", (toc11.getTime()-tic11.getTime()), "ms.");

var tic12 = new Date();
var promise = mandel.mandelbrot_async(640, 480, -0.5, 0.0, 4.0/640);
promise.then(function (res) {
	var toc12 = new Date();
	console.log("Mandelbrot set (640x480) calculated (async)!", "\tElapsed: ", (toc12.getTime()-tic12.getTime()), "ms.");
});

var tic13 = new Date();
mandel.writePGM("./mandelbrot.pgm", image);
var toc13 = new Date();
console.log("Mandelbrot set (640x480) written!", "\tElapsed: ", (toc13.getTime()-tic13.getTime()), "ms.");
console.log();

// wait for all pending promises
// TODO
console.log("------------------");
console.log("Done!");
